### ISSUE TITLE
Cannot upload/write to database

### ISSUE THREAD

**Issue by bertyterty** [dustin10/VichUploaderBundle#772]
Hi

I've been working on this for a few days now to absolutely no avail, and I'm starting to rip my hair out. I know I must be doing something wrong, but I just cannot see it.

So when I save my form with an image selected in the imageFile upload on the form, the page loads, and no errors appear, but nothing is saved into my database and nothing is stored on the filesystem. I've followed the docs as closely as I can, but still can't seem to get my head around it. My code is below, and I'm using Symfony 3.3.9

app/config/config.yml
```
vich_uploader:
    db_driver: orm

    mappings:
        inventory_image:
            uri_prefix:         /images/inventory
            upload_destination: '%kernel.root_dir%/../web/images/inventory'
```

src/AppBundle/Form/InventoryType.php
```<?php

namespace AppBundle\Form;

use Symfony\Component\Form\AbstractType;
use Symfony\Component\Form\FormBuilderInterface;
use Symfony\Component\Form\Extension\Core\Type\TextType;
use Symfony\Component\Form\Extension\Core\Type\HiddenType;
use Symfony\Component\Form\Extension\Core\Type\TextareaType;
use Symfony\Component\Form\Extension\Core\Type\NumberType;
use Symfony\Component\Form\Extension\Core\Type\SubmitType;
use Symfony\Component\Form\Extension\Core\Type\MoneyType;
use Symfony\Component\Form\Extension\Core\Type\CheckboxType;
use Symfony\Bridge\Doctrine\Form\Type\EntityType;
use Symfony\Component\Form\Extension\Core\Type\CollectionType;
use Symfony\Component\Form\Extension\Core\Type\FileType;
use Symfony\Component\OptionsResolver\OptionsResolver;
use Vich\UploaderBundle\Form\Type\VichImageType;

class InventoryType extends AbstractType
{
    public function buildForm(FormBuilderInterface $builder, array $options)
    {
        $builder
            ...
            ->add('imageFile', VichImageType::class)
            ...
    }

    public function configureOptions(OptionsResolver $resolver)
    {
        $resolver->setDefaults(array(
            'data_class' => 'AppBundle\Entity\Inventory',
        ));
    }
}
```

src/AppBundle/Entity/Inventory.php
```
<?php

namespace AppBundle\Entity;

use Symfony\Component\HttpFoundation\File\File;
use Symfony\Component\HttpFoundation\File\UploadedFile;
use Doctrine\Common\Collections\Collection,
Doctrine\Common\Collections\ArrayCollection;

/**
 * Inventory
 */
class Inventory
{
    ...
    /**
     * NOTE: This is not a mapped field of entity metadata, just a simple property.
     * 
     * @var File
     */
    private $imageFile;

    /**
     * @var string
     */
    private $imageName;

    /**
     * @var integer
     */
    private $imageSize;

    /**
     * @var \DateTime
     */
    private $updatedAt;
    ....
    /**
     * If manually uploading a file (i.e. not using Symfony Form) ensure an instance
     * of 'UploadedFile' is injected into this setter to trigger the  update. If this
     * bundle's configuration parameter 'inject_on_load' is set to 'true' this setter
     * must be able to accept an instance of 'File' as the bundle will inject one here
     * during Doctrine hydration.
     *
     * @param File|\Symfony\Component\HttpFoundation\File\UploadedFile $image
     *
     * @return Product
     */
    public function setImageFile(File $image = null)
    {
        $this->imageFile = $image;

        if ($image) {
            // It is required that at least one field changes if you are using doctrine
            // otherwise the event listeners won't be called and the file is lost
            $this->updatedAt = new \DateTime('now');
        }
        
        return $this;
    }

    /**
     * @return File|null
     */
    public function getImageFile()
    {
        return $this->imageFile;
    }

    /**
     * Set imageName
     *
     * @param string $imageName
     *
     * @return Inventory
     */
    public function setImageName($imageName)
    {
        $this->imageName = $imageName;

        return $this;
    }

    /**
     * Get imageName
     *
     * @return string
     */
    public function getImageName()
    {
        return $this->imageName;
    }

    /**
     * Set imageSize
     *
     * @param integer $imageSize
     *
     * @return Inventory
     */
    public function setImageSize($imageSize)
    {
        $this->imageSize = $imageSize;

        return $this;
    }

    /**
     * Get imageSize
     *
     * @return integer
     */
    public function getImageSize()
    {
        return $this->imageSize;
    }

    /**
     * Set updatedAt
     *
     * @param \DateTime $updatedAt
     *
     * @return Inventory
     */
    public function setUpdatedAt($updatedAt)
    {
        $this->updatedAt = $updatedAt;

        return $this;
    }

    /**
     * Get updatedAt
     *
     * @return \DateTime
     */
    public function getUpdatedAt()
    {
        return $this->updatedAt;
    }
    ....
}
```

src/AppBundle/Resources/vich_uploader/entity.inventory.yml
```
AppBundle\Entity\Inventory:
    imageFile:
        mapping:           inventory_image # required
        filename_property: imageName     # required
        size:              imageSize
```

src/AppBundle/Resources/doctrine.orm.yml
```
AppBundle\Entity\Inventory:
    type: entity
    table: null
    repositoryClass: AppBundle\Repository\InventoryRepository
    id:
        id:
            type: integer
            id: true
            generator:
                strategy: AUTO
    fields:
    ....
    imageName:
            type: string
            length: '255'
            nullable: true
        imageSize:
            type: integer
            nullable: true
        updatedAt:
            type: datetime
            nullable: true
    ....
```

I use yml for all my mapping info, but I have tried using annotations too but that had no effect. I know that Vich is reading the yml config file because I did have my mapping set incorrectly in there and I got a 'Mapping not found' error which when I updated the Vich config yml file, fixed it.

Any help really appreciated, and let me know if you want me to send any more information. A little background to what I am trying to achieve is my system is going to be used to track inventory we have on hire stock amongst other things, and I want to be able to upload a picture of that stock item to be used in the list view, so it is literally one picture, and once I get the upload actually working, the files will be named by the unique ID of the inventory, as any new uploaded files I will want to overwrite the old one if somebody wanted to update that image.

Thank you for any help you can provide, and sorry if it's something blindingly obvious staring me in the face

**Comment by garak** [comment]
Can you try to use the debug commands and see if your entity is actually mapped?

**Comment by bertyterty** [comment]
Thanks for replying garak!

I have run the following commands:
- `php bin/console vich:mapping:debug inventory_image`
- `php bin/console vich:mapping:list-classes`
- `php bin/console vich:mapping:debug-class AppBundle\\Entity\\Inventory`

vich:mapping:debug inventory_image produced the following response:
```
Andrews-MacBook-Pro:scope andrew$ php bin/console vich:mapping:debug inventory_image
Debug information for mapping inventory_image
uri_prefix: '/images/inventory'
upload_destination: '/Users/andrew/Documents/GitHub/scope/app/../web/images/inventory'
namer: array (
  'service' => NULL,
  'options' => NULL,
)
directory_namer: array (
  'service' => NULL,
  'options' => NULL,
)
delete_on_remove: true
delete_on_update: true
inject_on_load: false
db_driver: 'orm'
```

vich:mapping:list-classes returned:
```
Andrews-MacBook-Pro:scope andrew$ php bin/console vich:mapping:list-classes
Looking for uploadable classes.
Found AppBundle\entity\inventory
Found 1 classes.
NOTE: Only classes configured using XML or YAML are displayed.
```

vich:mapping:debug-class AppBundle\\\Entity\\\Inventory returned:
```
Andrews-MacBook-Pro:scope andrew$ php bin/console vich:mapping:debug-class AppBundle\\Entity\\Inventory
Introspecting class AppBundle\Entity\Inventory:
Found field "imageFile", storing file name in "imageName" and using mapping "inventory_image"
```

I did note that the list-classes command returned AppBundle\entity\inventory, but to debug a class, I had to use AppBundle\\\Entity\\\Inventory. I understand having to escape the backslashes, but I'm not sure why the case had to change. When I tried it lowercase it returned class not found, but presumably the list-classes command should return it in the correct case...

That aside, as far as I can tell from there my mapping is correct. I haven't set a namer yet, although I do intend to, however the documentation doesn't call for this as a basic step so I am assuming that that is not what is causing the problem.

Thanks for your help :)

**Comment by garak** [comment]
Just to be clear: when you say "nothing is saved into my database", do you mean that persisting your entity does not work at all? If so, the problem is likely outside this bundle.

**Comment by bertyterty** [comment]
Apologies, no. The rest of the entity is persisted.

I did suspect maybe I had an issue with the setImageFile method in the entity where it sets the updatedAt value as well to ensure doctrine persists the entity if only the file changes. The rest of the inventory does persist properly. I have tried selecting an image file to upload whilst at the same time changing a value in another attribute on the form and thof n hitting save, and the attribute updated but the image still did not, and the columns in the database relating to the image remained at null

**Comment by garak** [comment]
Try to `dump($image)` inside `setImageFile()` method

**Comment by bertyterty** [comment]
I'm not in the office at the moment, but I did try that a few days ago and it actually didn't run. I've changed quite a few things since then so I'll try it again when I'm back at my computer in an hour or two

**Comment by bertyterty** [comment]
Ok, so I tried the `dump($image);` in the setImageFile method, and again nothing. Just to make sure, I dumped another attribute in the setter method and that one returned as I would have expected, so for whatever reason it seems the setImageFile method isn't being called, so I presume that's the problem, but I have no idea why. I've only been working with Symfony for around 6 months, but my understanding is that the attributes linked to the form should call the getters and setters when the form is rendered and submitted, as part of the default functionality.

**Comment by garak** [comment]
It looks like there's some problem before VichUploaderBundle is involved.
Try to use a simple FileType and see if the file is passes to `setImageFile` method

**Comment by bertyterty** [comment]
You were right!

As I suspected, I was doing something ridiculously stupid!

It turns out that I had rendered the imageFile field as a form, rather than form_row :(
```{{ form(form.imageFile) }}```
vs
```{{ form_row(form.imageFile) }}```

It's working now, entity is persisted, image is in the correct folder, if only I had paid more attention to the template I would have found it long ago!

Thanks a lot garak for giving me your time in helping me :)
